dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -113,7 +139,27 @@ public Object getNextEvent(String handlerId) throws InterruptedException {
 	}
 
 	public void addHandler(String handlerId) {
-		handlerEventMap.putIfAbsent(handlerId, new HandlerStats(handlerId, sourceId));
+		RERWLOCK.writeLock().lock();
+		try {
+			//If it is first async handler subscribed, then create the main buffer
+			if(ringBuffer == null) {
+				ringBuffer = new Buffer<Object>(capacity);
+			}
+			/*
+			 * Every new Asynchronous handler starts off with all events from EMQ.
+			 * So we create a copy of current EMQ and sends those messages to RingBuffer
+			 */
+			if(earlyMessageQueue != null && !earlyMessageQueue.isEmpty()) {
+				Object holder[] = new Object[EARLY_MESSAGE_QUEUE_SIZE];
+				Object[] messagesList = earlyMessageQueue.toArray(holder);
```

Comment:
it doesn't make sense to copy the EMQ here since you need to avoid anyone adding to the EMQ while you're doing the copy to the ringBuffer anyway. You can just copy straight from the EMQ into the ringBuffer.